### PR TITLE
Handle the upgrade-series now there are no unit agents.

### PR DESCRIPTION
**Pull request by howbazaar**
This branch updates the service interface to deal with the lack of unit agents.

The deployer when starting the unit workers now ensures that the tools directory is correct.

## QA steps

Start with a new model where we deploy some units.

```sh
juju add-model upgrade
juju deploy ./acceptancetests/repository/charms/dummy-sink --series xenial
juju deploy ./acceptancetests/repository/charms/dummy-subordinate --series xenial
watch -c juju status --color
```
wait until they are fully deployed and ready, then run the following to start the upgrade
```sh
juju upgrade-series 0 prepare bionic
juju run --machine 0 'sudo shutdown -r now'
```
The reboot is needed because do-release-upgrade requires that the machine has been rebooted since it was created.
```sh
juju ssh 0
sudo do-release-upgrade -f DistUpgradeViewNonInteractive
sudo reboot
```
Wait for the machine to come back up and show as connected in status, then run the final command.
```sh
juju upgrade-series 0 complete
```

### CODE REVIEW

File: worker/upgradeseries/upgrader_test.go [hunk 1]
```diff
@@ -44,7 +44,7 @@ func (s *upgraderSuite) TestNotToSystemdCopyToolsOnly(c *gc.C) {
 
 	// No systemd file changes; just the new tools for the target series.
 	s.manager.EXPECT().CopyAgentBinary(
-		s.machineService, s.unitServices, paths.NixDataDir, "trusty", "precise", version.Current,
```
Comment (manadart) [inline worker/upgradeseries/upgrader_test.go:47]: Might as well remove the suite member and setup for this.

File: worker/upgradeseries/upgrader.go [hunk 2]
```diff
@@ -114,17 +114,9 @@ func (u *upgrader) ensureSystemdFiles() error {
...
-		u.logger.Infof("agents written and linked by systemd: %s", strings.Join(services, ", "))
-	}
-	if len(links) > 0 {
-		u.logger.Infof("agents written and linked by symlink: %s", strings.Join(links, ", "))
-	}
-
-	return errors.Annotatef(err, "failed to write agents: %s", strings.Join(failed, ", "))
+	return errors.Annotatef(
+		u.manager.WriteSystemdAgents(u.machineAgent, paths.NixDataDir, systemdMultiUserDir),
+		"failed to write machine agent")
```
Comment (manadart) [inline worker/upgradeseries/upgrader.go:119]: I always annotate these as "writing machine agent", the "failed" part being implied by the non-nil error.

File: worker/upgradeseries/upgrader.go [hunk 3]
```diff
@@ -114,17 +114,9 @@ func (u *upgrader) ensureSystemdFiles() error {
...
-	if len(services) > 0 {
-		u.logger.Infof("agents written and linked by systemd: %s", strings.Join(services, ", "))
-	}
-	if len(links) > 0 {
-		u.logger.Infof("agents written and linked by symlink: %s", strings.Join(links, ", "))
-	}
-
-	return errors.Annotatef(err, "failed to write agents: %s", strings.Join(failed, ", "))
+	return errors.Annotatef(
+		u.manager.WriteSystemdAgents(u.machineAgent, paths.NixDataDir, systemdMultiUserDir),
```
Comment (manadart) [inline worker/upgradeseries/upgrader.go:118]: We should for clarity, change all of these pluralities to singulars now that we are only working with the machine agent.

File: service/agentconf.go [hunk 4]
```diff
@@ -41,22 +41,22 @@ type SystemdServiceManager interface {
...
-	// CopyAgentBinary copies all the tools into the path specified for each agent.
+	// CopyAgentBinary copies the tools into the path specified for the machine agent.
 	CopyAgentBinary(
-		machineAgent string, unitAgents []string, dataDir, toSeries, fromSeries string, jujuVersion version.Number,
+		machineAgent string, dataDir, toSeries, fromSeries string, jujuVersion version.Number,
 	) error
 
 	// StartAllAgents starts all the agents in the machine with specified series.
-	StartAllAgents(machineAgent string, unitAgents []string, dataDir string) (string, []string, error)
+	StartAllAgents(machineAgent string, dataDir string) (string, error)
```
Comment (manadart) [inline service/agentconf.go:59]: Looks like this is never called except in tests. Might as well turf it.

File: service/agentconf.go [hunk 5]
```diff
@@ -94,30 +94,19 @@ func NewServiceManager(
...
 // /etc/systemd/system path.
 func (s *systemdServiceManager) WriteServiceFiles() error {
-	machineAgent, unitAgents, _, err := s.FindAgents(paths.NixDataDir)
+	machineAgent, _, _, err := s.FindAgents(paths.NixDataDir)
 	if err != nil {
 		return errors.Trace(err)
 	}
 
-	serviceNames, linkNames, failed, err := s.WriteSystemdAgents(
+	err = s.WriteSystemdAgents(
```
Comment (manadart) [inline service/agentconf.go:102]: Singular for all these now.

File: service/agentconf.go [hunk 6]
```diff
@@ -94,30 +94,19 @@ func NewServiceManager(
 // WriteServiceFiles writes service files to the standard
 // /etc/systemd/system path.
 func (s *systemdServiceManager) WriteServiceFiles() error {
-	machineAgent, unitAgents, _, err := s.FindAgents(paths.NixDataDir)
+	machineAgent, _, _, err := s.FindAgents(paths.NixDataDir)
```
Comment (manadart) [inline service/agentconf.go:97]: Maybe a note on `FindAgents` about retention of the unit discovery for upgrade steps?

File: worker/upgradeseries/upgrader_test.go (hunk 1 above)
Comment (howbazaar) [reply worker/upgradeseries/upgrader_test.go:47]: OK

File: worker/upgradeseries/upgrader.go (hunk 2 above)
Comment (howbazaar) [reply worker/upgradeseries/upgrader.go:119]: Sounds like a good idea.

File: service/agentconf.go (hunk 4 above)
Comment (howbazaar) [reply service/agentconf.go:59]: Turfing.

File: service/agentconf.go (hunk 5 above)
Comment (howbazaar) [reply service/agentconf.go:102]: Done

File: worker/upgradeseries/upgrader.go (hunk 3 above)
Comment (howbazaar) [reply worker/upgradeseries/upgrader.go:118]: Fair enough.

File: service/agentconf.go (hunk 6 above)
Comment (howbazaar) [reply service/agentconf.go:97]: ok.

**Comment by howbazaar** [comment]
!!build!!

**Comment by howbazaar** [comment]
$$merge$$

**Pull request merged**